File: SampleApps/SampleAppObjC/advanced/HistoryViewController.m
//
//  HistoryViewController.m
//  SampleAppObjC
//
//  Copyright © 2018 BEAR SAS. All rights reserved.
//

#import "HistoryViewController.h"
#import "NoStoryboardViewController.h"
#import "UserDefaultKeys.h"
#import <BearSDK/BearSDK.h>

@interface HistoryViewController ()

@property (nonatomic, nonnull, strong, readonly) NSArray<NSNumber*>* recognizedMarkers;
@property (weak, nonatomic, nullable) IBOutlet UIBarButtonItem *unloadBtn;
@property (weak, nonatomic) IBOutlet UIBarButtonItem *dismissBtn;
@property (weak, nonatomic) IBOutlet UIToolbar *toolbar;

@end

@implementation HistoryViewController

static NSString* reuseIdentifier = @"historyReuseId";

- (void)viewDidLoad {
    [super viewDidLoad];
    
    _recognizedMarkers = [[NSUserDefaults standardUserDefaults] arrayForKey:UserDefaultKeys.recognizedMarkers];
    
    if (_hideUnloadBtn) {
        NSMutableArray* copyArr = [_toolbar.items mutableCopy];
        [copyArr removeObjectAtIndex:0];
        _toolbar.items = copyArr;
    }
    
    if (_recognizedMarkers.count == 0) {
        UILabel* lbl = [[UILabel alloc] init];
        lbl.text = @"Scanned markers will appear here.\nYou can scan using any of samples";
        lbl.numberOfLines = 2;
        [lbl sizeToFit];
        [self.view addSubview:lbl];
        lbl.center = self.view.center;
    }
}

-(void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    if (_unloadBtn)
        _unloadBtn.enabled = BearSDK.shared.isLoaded;
    
    if ([self navigationController]) {
        [[self navigationController] setNavigationBarHidden:NO animated:animated];
        [_dismissBtn setEnabled:NO];
    }
}

- (IBAction)unloadTapped {
    if (!_unloadBtn || !BearSDK.shared.isLoaded) {
        return NSLog(@"BearSDK is not loaded, nothing to destroy");
    }
    [BearSDK.shared releaseResources];
    _unloadBtn.enabled = false;
}

- (IBAction)dismiss {
    [self dismissViewControllerAnimated:YES completion:nil];
}

// MARK: UITableViewDataSource

-(NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return _recognizedMarkers.count;
}

-(UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell* cell = [tableView dequeueReusableCellWithIdentifier:reuseIdentifier forIndexPath:indexPath];
    cell.textLabel.text = [NSString stringWithFormat:@"%li", (long)_recognizedMarkers[indexPath.row].integerValue];
    return cell;
}

// MARK: UITableViewDelegate

-(void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    NoStoryboardViewController* vc = [[NoStoryboardViewController alloc] init];
    vc.modalPresentationStyle = UIModalPresentationFullScreen;
    [vc.handler showARSceneWithoutTrackingWithMarkerId:_recognizedMarkers[indexPath.row].integerValue];
    [self showViewController:vc sender:self];
}

@end

File: SampleApps/SampleApp/advanced/HistoryViewController.swift
//
//  HistoryViewController.swift
//  SampleApp
//
//  Copyright © 2018 BEAR SAS. All rights reserved.
//

import UIKit
import BearSDK

final class HistoryViewController: UIViewController {

    @IBOutlet weak var dismissBtn: UIBarButtonItem!
    @IBOutlet weak var unloadBtn: UIBarButtonItem?
    @IBOutlet weak var toolbar: UIToolbar!
    
    private static let reuseIdentifier = "historyReuseId"
    private let recognizedMarkers = UserDefaults.standard.array(forKey: UserDefaultKeys.recognizedMarkers) as? [Int] ?? []
    var hideUnloadBtn = false
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        if hideUnloadBtn {
            toolbar.items?.removeFirst()
        }
        
        if recognizedMarkers.isEmpty {
            let label = UILabel()
            label.text = "Scanned markers will appear here.\nYou can scan using any of samples"
            label.numberOfLines = 2
            label.sizeToFit()
            view.addSubview(label)
            label.center = view.center
        }
    }
    
    @IBAction func unloadTapped() {
        guard let destroy = unloadBtn, BearSDK.shared.isLoaded else {
            return print("BearSDK is not loaded, nothing to destroy")
        }
        BearSDK.shared.releaseResources()
        destroy.isEnabled = false
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        unloadBtn?.isEnabled = BearSDK.shared.isLoaded
        if let nav = navigationController {
            nav.setNavigationBarHidden(false, animated: animated)
            dismissBtn.isEnabled = false
        }
    }
    
    @IBAction func dismiss() {
        dismiss(animated: true, completion: nil)
    }
}

extension HistoryViewController: UITableViewDataSource {
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        recognizedMarkers.count
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: HistoryViewController.reuseIdentifier, for: indexPath)
        cell.textLabel?.text = "\(recognizedMarkers[indexPath.row])"
        return cell
    }
}

extension HistoryViewController: UITableViewDelegate {
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let vc = NoStoryboardViewController()
        vc.modalPresentationStyle = .fullScreen
        vc.handler.showARSceneWithoutTracking(withMarkerId: recognizedMarkers[indexPath.row])
        show(vc, sender: self)
    }
}
